File: agentic_os/mission/operator_sdk.py
"""Operator SDK — how an agentic app becomes a capability operator (a v5 "process").

An app declares its capabilities (syscalls) once and gets, for free:
  * `GET /capabilities`  — the manifest the Mission Planner discovers + reasons over;
  * `POST /invoke`       — run a named capability with server-side idempotency dedupe
                           (exactly-once for side effects, keyed on the Idempotency-Key);
  * a `LocalOperatorClient` so the control plane can drive co-located operators in-process
    (no HTTP hop) while remote apps are driven by `HTTPOperatorClient`.

This is the seam that turns `agents/<app>` (today: an AgentConsole) into a Mission Runtime
operator without changing how the app runs standalone.
"""

from dataclasses import dataclass, field
from typing import Callable

from .types import CapabilityManifest, CapabilitySpec, NodeCost, to_jsonable

Handler = Callable[[dict], dict]


@dataclass
class Capability:
    """A capability spec bound to its handler."""
    spec: CapabilitySpec
    handler: Handler


def capability(name: str, handler: Handler, *, provides: list[str] | None = None,
               operator: str = "", inputs: dict[str, str] | None = None,
               outputs: dict[str, str] | None = None, side_effecting: bool = False,
               approval_required: bool = False, undo: str | None = None,
               permissions: list[str] | None = None, deterministic: bool = True,
               estimated_value: str = "medium", usd: float = 0.0, latency_ms: int = 0,
               required_authority: list[str] | None = None, isolation_class: str = "",
               network: list[str] | None = None,
               data_classifications: list[str] | None = None) -> Capability:
    """Ergonomic builder for one capability + its handler.

    The security fields (v0.3.x, all optional) are the metadata the opt-in Executor security seams read:
    `required_authority` feeds the delegated-authority gate (`authority_for`), `isolation_class` the
    sandbox seam (`isolation_for`), and `network`/`data_classifications` the SecurityMonitor's
    boundary telemetry. Declaring nothing leaves the capability unchanged."""
    spec = CapabilitySpec(
        name=name, operator=operator, provides=provides or [], inputs=inputs or {},
        outputs=outputs or {}, side_effecting=side_effecting, approval_required=approval_required,
        undo=undo, permissions=permissions or [], deterministic=deterministic,
        estimated_value=estimated_value, cost=NodeCost(usd=usd, latency_ms=latency_ms),
        required_authority=required_authority or [], isolation_class=isolation_class,
        network=network or [], data_classifications=data_classifications or [],
    )
    return Capability(spec=spec, handler=handler)


class Operator:
    """One app's capability operator: manifest + handlers + idempotency dedupe."""

    def __init__(self, name: str, capabilities: list[Capability]):
        self.name = name
        for c in capabilities:
            c.spec.operator = c.spec.operator or name
        self.manifest = CapabilityManifest(operator=name, capabilities=[c.spec for c in capabilities])
        self._handlers: dict[str, Handler] = {c.spec.name: c.handler for c in capabilities}
        self._seen: dict[str, dict] = {}       # idempotency_key -> first result (exactly-once)
        self.calls: list[tuple[str, str]] = []  # (capability, idempotency_key) — for assertions

    def invoke(self, capability: str, inputs: dict, idempotency_key: str = "") -> dict:
        if idempotency_key and idempotency_key in self._seen:
            return self._seen[idempotency_key]
        fn = self._handlers.get(capability)
        if fn is None:
            raise KeyError(f"operator '{self.name}' has no capability '{capability}'")
        self.calls.append((capability, idempotency_key))
        result = fn(inputs) or {}
        if idempotency_key:
            self._seen[idempotency_key] = result
        return result

    # ── HTTP surface (mount into the app's FastAPI) ──────────────────────────
    def router(self):
        from fastapi import APIRouter, Header, HTTPException
        from pydantic import BaseModel

        router = APIRouter(tags=[f"operator:{self.name}"])

        class InvokeBody(BaseModel):
            capability: str
            inputs: dict = {}
            idempotency_key: str = ""
            mission_id: str | None = None
            node_id: str | None = None

        @router.get("/capabilities")
        def capabilities():
            return to_jsonable(self.manifest)

        @router.post("/invoke")
        def invoke(body: InvokeBody, idempotency_key: str | None = Header(default=None)):
            key = body.idempotency_key or idempotency_key or ""
            try:
                return {"result": self.invoke(body.capability, body.inputs, key)}
            except KeyError as e:
                raise HTTPException(404, str(e))

        return router


class LocalOperatorClient:
    """An `OperatorClient` over in-process Operator objects — the control plane driving
    co-located operators without an HTTP hop. Remote apps use HTTPOperatorClient instead."""

    def __init__(self, operators: dict[str, Operator]):
        self._ops = operators

    def invoke(self, operator: str, capability: str, inputs: dict, idempotency_key: str) -> dict:
        op = self._ops.get(operator)
        if op is None:
            raise KeyError(f"no operator '{operator}' registered")
        return op.invoke(capability, inputs, idempotency_key)

File: go/mission/operator_sdk.go
package mission

// Operator SDK — how an agentic app becomes a capability operator (a v5 "process"). An app declares
// its capabilities once and gets GET /capabilities (the manifest the planner discovers) + POST
// /invoke (run a capability, server-side idempotency dedupe) + a LocalOperatorClient for co-located
// operators. This is the seam that turns agents/<app> into a Mission Runtime operator without
// changing how it runs standalone. Go port of agentic_os/mission/operator_sdk.py.

import (
	"encoding/json"
	"fmt"
	"net/http"
)

// Capability is a capability spec bound to its handler.
type Capability struct {
	Spec    CapabilitySpec
	Handler Handler
}

// CapOption configures a capability spec (the ergonomic builder args).
type CapOption func(*CapabilitySpec)

func WithProvides(p ...string) CapOption       { return func(s *CapabilitySpec) { s.Provides = p } }
func WithPermissions(p ...string) CapOption    { return func(s *CapabilitySpec) { s.Permissions = p } }
func WithInputs(m map[string]string) CapOption { return func(s *CapabilitySpec) { s.Inputs = m } }
func WithOutputs(m map[string]string) CapOption {
	return func(s *CapabilitySpec) { s.Outputs = m }
}
func SideEffecting() CapOption        { return func(s *CapabilitySpec) { s.SideEffecting = true } }
func ApprovalRequired() CapOption     { return func(s *CapabilitySpec) { s.ApprovalRequired = true } }
func WithUndo(u string) CapOption     { return func(s *CapabilitySpec) { s.Undo = u } }
func WithValue(v string) CapOption    { return func(s *CapabilitySpec) { s.EstimatedValue = v } }
func WithOperator(o string) CapOption { return func(s *CapabilitySpec) { s.Operator = o } }
func WithConcurrencyMode(m string) CapOption { return func(s *CapabilitySpec) { s.ConcurrencyMode = m } }
func WithConcurrencyKey(k string) CapOption  { return func(s *CapabilitySpec) { s.ConcurrencyKey = k } }
func WithResourceKeys(k ...string) CapOption { return func(s *CapabilitySpec) { s.ResourceKeys = k } }
func WithMaxParallelism(n int) CapOption     { return func(s *CapabilitySpec) { s.MaxParallelism = n } }
func WithCost(usd float64, latencyMs int) CapOption {
	return func(s *CapabilitySpec) { s.Cost = NodeCost{USD: usd, LatencyMs: latencyMs} }
}

// NewCapability is the ergonomic builder for one capability + its handler.
func NewCapability(name string, handler Handler, opts ...CapOption) Capability {
	spec := NewCapabilitySpec(name, "")
	for _, o := range opts {
		o(&spec)
	}
	return Capability{Spec: spec, Handler: handler}
}

// Operator is one app's capability operator: manifest + handlers + idempotency dedupe.
type Operator struct {
	Name     string
	Manifest CapabilityManifest
	handlers map[string]Handler
	seen     map[string]map[string]any
	Calls    [][2]string // (capability, idempotency_key) — for assertions
}

func NewOperator(name string, caps []Capability) *Operator {
	specs := make([]CapabilitySpec, 0, len(caps))
	handlers := make(map[string]Handler, len(caps))
	for i := range caps {
		if caps[i].Spec.Operator == "" {
			caps[i].Spec.Operator = name
		}
		specs = append(specs, caps[i].Spec)
		handlers[caps[i].Spec.Name] = caps[i].Handler
	}
	return &Operator{
		Name: name, Manifest: CapabilityManifest{Operator: name, Capabilities: specs},
		handlers: handlers, seen: map[string]map[string]any{},
	}
}

func (o *Operator) Invoke(capability string, inputs map[string]any, idempotencyKey string) (map[string]any, error) {
	if idempotencyKey != "" {
		if r, ok := o.seen[idempotencyKey]; ok {
			return r, nil
		}
	}
	fn, ok := o.handlers[capability]
	if !ok {
		return nil, fmt.Errorf("operator %q has no capability %q", o.Name, capability)
	}
	o.Calls = append(o.Calls, [2]string{capability, idempotencyKey})
	result, err := fn(inputs)
	if err != nil {
		return nil, err
	}
	if result == nil {
		result = map[string]any{}
	}
	if idempotencyKey != "" {
		o.seen[idempotencyKey] = result
	}
	return result, nil
}

// Handler returns the operator's HTTP surface (mount into the app's server): GET /capabilities and
// POST /invoke with server-side idempotency (also honouring the Idempotency-Key header).
func (o *Operator) HTTPHandler() http.Handler {
	mux := http.NewServeMux()
	mux.HandleFunc("GET /capabilities", func(w http.ResponseWriter, _ *http.Request) {
		writeJSON(w, http.StatusOK, jsonable(o.Manifest))
	})
	mux.HandleFunc("POST /invoke", func(w http.ResponseWriter, r *http.Request) {
		var body struct {
			Capability     string         `json:"capability"`
			Inputs         map[string]any `json:"inputs"`
			IdempotencyKey string         `json:"idempotency_key"`
			MissionID      string         `json:"mission_id"`
			NodeID         string         `json:"node_id"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]any{"detail": err.Error()})
			return
		}
		key := body.IdempotencyKey
		if key == "" {
			key = r.Header.Get("Idempotency-Key")
		}
		res, err := o.Invoke(body.Capability, body.Inputs, key)
		if err != nil {
			writeJSON(w, http.StatusNotFound, map[string]any{"detail": err.Error()})
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"result": res})
	})
	return mux
}

func writeJSON(w http.ResponseWriter, code int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	_ = json.NewEncoder(w).Encode(v)
}

// LocalOperatorClient drives co-located in-process Operators (no HTTP hop). Remote apps use
// HTTPOperatorClient instead.
type LocalOperatorClient struct{ ops map[string]*Operator }

func NewLocalOperatorClient(ops map[string]*Operator) *LocalOperatorClient {
	return &LocalOperatorClient{ops: ops}
}

func (c *LocalOperatorClient) Invoke(operator, capability string, inputs map[string]any, idempotencyKey string) (map[string]any, error) {
	op, ok := c.ops[operator]
	if !ok {
		return nil, fmt.Errorf("no operator %q registered", operator)
	}
	return op.Invoke(capability, inputs, idempotencyKey)
}
